### PR TITLE
RavenDB-18420: Reschedule Backup Timer If DocumentDatabase Failed To Load

### PR DESCRIPTION
**Pull request by garayx**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18420

### Type of change

- Bug fix

### How risky is the change?

- Low 
### Backward compatibility

- Not relevant

### Testing 

- Tests have been added that prove the fix is effective or that the feature works
 - Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- It has been verified by manual testing

### Is there any existing behavior change of other features due to this change?
- No

### CODE REVIEW

File: src/Raven.Server/Documents/DatabasesLandlord.cs [hunk 1]
```diff
@@ -1039,7 +1041,23 @@ private void StartDatabaseOnTimer(string name, DateTime? wakeup = null)
...
+                    var t = TryGetOrCreateResourceStore(name, wakeup);
+
+                    t.ContinueWith(async _ =>
+                    {
+                        try
+                        {
+                            await t;
+                        }
+                        catch (DatabaseConcurrentLoadTimeoutException)
+                        {
```
Comment (grisha-kotler) [inline src/Raven.Server/Documents/DatabasesLandlord.cs]: Can you log this?

**Review by grisha-kotler — APPROVED** [review]

File: src/Raven.Server/Documents/DatabasesLandlord.cs (hunk 1 above)
Comment (garayx) [reply src/Raven.Server/Documents/DatabasesLandlord.cs]: done

File: src/Raven.Server/Documents/DatabasesLandlord.cs [hunk 2]
```diff
@@ -79,6 +80,7 @@ internal class TestingStuff
             internal int? HoldDocumentDatabaseCreation = null;
             internal bool PreventedRehabOfIdleDatabase = false;
             internal Action<DocumentDatabase> OnBeforeDocumentDatabaseInitialization;
+            internal ManualResetEventSlim? RescheduleDatabaseWakeupMRE = null;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/DatabasesLandlord.cs]: remove `nullable`,  just call it `RescheduleDatabaseWakeupMre`

File: src/Raven.Server/Documents/DatabasesLandlord.cs [hunk 3]
```diff
@@ -1039,7 +1041,26 @@ private void StartDatabaseOnTimer(string name, DateTime? wakeup = null)
...
+                            // database failed to load, retry after 1 min
+                            if (ForTestingPurposes?.RescheduleDatabaseWakeupMRE != null)
+                                ForTestingPurposes?.RescheduleDatabaseWakeupMRE.Set();
+
+                            if (_logger.IsInfoEnabled)
+                                await _logger.InfoAsync($"Failed to start database '{name}' on timer, will retry the wakeup in '{_dueTimeOnRetry}' ms", e);
+
+                            RescheduleDatabaseWakeup(name, milliseconds: _dueTimeOnRetry, wakeup);
+                        }
+                    });
```
Comment (ppekrol) [inline src/Raven.Server/Documents/DatabasesLandlord.cs]: OnlyOnRanToCompletion maybe?

File: src/Raven.Server/Documents/DatabasesLandlord.cs [hunk 4]
```diff
@@ -1039,7 +1041,26 @@ private void StartDatabaseOnTimer(string name, DateTime? wakeup = null)
...
-                    TryGetOrCreateResourceStore(name, wakeup);
+                    var t = TryGetOrCreateResourceStore(name, wakeup);
+
+                    t.ContinueWith(async _ =>
+                    {
+                        try
+                        {
+                            await t;
+                        }
+                        catch (DatabaseConcurrentLoadTimeoutException e)
```
Comment (ppekrol) [inline src/Raven.Server/Documents/DatabasesLandlord.cs]: what will happen if other exception will be thrown here? Unobserved task ex?

**Review by ppekrol — APPROVED** [review]

File: src/Raven.Server/Documents/DatabasesLandlord.cs [hunk 5]
```diff
@@ -1039,7 +1041,21 @@ private void StartDatabaseOnTimer(string name, DateTime? wakeup = null)
...
 
-                    TryGetOrCreateResourceStore(name, wakeup);
+                    _ = TryGetOrCreateResourceStore(name, wakeup).ContinueWith(t =>
+                          {
+                              var ex = t.Exception.ExtractSingleInnerException();
+                              if (ex is DatabaseConcurrentLoadTimeoutException e)
+                              {
+                                  // database failed to load, retry after 1 min
+                                  if (ForTestingPurposes?.RescheduleDatabaseWakeupMre != null)
+                                      ForTestingPurposes?.RescheduleDatabaseWakeupMre.Set();
```
Comment (ayende) [inline src/Raven.Server/Documents/DatabasesLandlord.cs]:
```suggestion
                                      ForTestingPurposes?.RescheduleDatabaseWakeupMre?.Set();
```

**Review by ayende — APPROVED** [review]
Very minor comment

File: src/Raven.Server/Documents/DatabasesLandlord.cs (hunk 5 above)
Comment (garayx) [reply src/Raven.Server/Documents/DatabasesLandlord.cs]: done

**Pull request merged**